configuration/pih/liquibase/sql/void_duplicate_registrations.sql

```diff
@@ -0,0 +1,158 @@
+
+-- start checking for dupes -----------------------------------
+DROP TEMPORARY TABLE IF EXISTS temp_dup_info;
+CREATE TEMPORARY TABLE  temp_dup_info
+(group_id int(11)  NOT NULL AUTO_INCREMENT,           
+mothers_first_name                VARCHAR(255),  
+address_level_1                   VARCHAR(255),  
+address_level_4                   VARCHAR(255),  
+address_level_2                          VARCHAR(255),  
+address_level_5                           VARCHAR(255),  
+address_level_3                           VARCHAR(255),  
+telephone_number                  VARCHAR(255),  
+reg_location                      varchar(50),   
+registration_date                 date,          
+registration_entry_date           datetime,      
+creator                           int(11),       
+user_entered                      varchar(50),   
+name                              varchar(50),   
+family_name                       varchar(50),   
+dob                               date,          
+gender                            varchar(2),    
+PRIMARY KEY (group_id)
+);
+
+-- load table with registration info of all the duplicates, using group_id to tie together the groups of dups 
+INSERT INTO temp_dup_info(name, family_name, dob, gender, mothers_first_name, telephone_number, address_level_1, address_level_2, address_level_3, address_level_4, address_level_5, reg_location, registration_date, registration_entry_date, user_entered)
+ select name, family_name, dob, gender, mothers_first_name, telephone_number, address_level_1, address_level_2, address_level_3, address_level_4, address_level_5, reg_location, registration_date, DATE(registration_entry_date), user_entered
+ from temp_patients
+ group by name, family_name, dob, gender, mothers_first_name, telephone_number, address_level_1, address_level_2, address_level_3, address_level_4, address_level_5, reg_location, registration_date, DATE(registration_entry_date), user_entered
+ having count(*) > 1
+ ;
+
+-- join together the dup information with the entire list of patients
+-- to create a list of patients with patient_ids grouped together
+DROP TEMPORARY TABLE IF EXISTS temp_potential_dups;
+CREATE TEMPORARY TABLE  temp_potential_dups
+select d.group_id, p.* from temp_patients p
+inner join temp_dup_info d
+on d.name <=> p.name -- -- note '<=>' will compare nulls properly
+and d.family_name <=> p.family_name
+and d.dob <=> p.dob
+and d.gender <=> p.gender
+and d.mothers_first_name <=> p.mothers_first_name
+and d.telephone_number <=> p.telephone_number
+and d.address_level_1 <=> p.address_level_1
+and d.address_level_2 <=> p.address_level_2
+and d.address_level_3 <=> p.address_level_3
+and d.address_level_4 <=> p.address_level_4
+and d.address_level_5 <=> p.address_level_5
+and d.reg_location <=> p.reg_location
+and d.registration_date <=> p.registration_date
+and date(d.registration_entry_date) <=> date(p.registration_entry_date)
+and d.user_entered <=> p.user_entered;
```

Comment:
I'm probably missing this, but why do we have to do a comparison on all the fields again?  Doesn't the group part handle that?

Reply:
This is the part that, once you identify all the duplicate rows without the patient_id, it joins these rows back to the original dataset (that HAS patient_id) so in order to join to the correct row, you need to make sure every field matches.  

I know there more straightforward ways to do it, but after lots of trial and error, by far the fastest way to do it is to group by all the fields you want to compare first.